jvb-api/jvb-api-client/src/main/kotlin/org/jitsi/videobridge/api/util/SynchronousWebSocketClient.kt

```diff
@@ -0,0 +1,130 @@
+ *
+ * All messages are sent as [Frame.Text], every message *must* have a response
+ * from the server[1], and all responses *must* be of type [Frame.Text].
+ *
+ * [1]: Rather than implement a request-response-correlation scheme, this
+ * client simply assumes all requests will have a response and that a request
+ * will be responded to before the next request is handled.  With this
+ * assumption, we can merely handle a response each time we send a request
+ * and it will keep requests and responses 'aligned'.
+ */
+class SynchronousWebSocketClient(
+    private val client: HttpClient,
+    private val host: String,
+    private val port: Int,
+    /**
+     * The path of the remote websocket URL
+     */
+    private val path: String,
+    parentLogger: Logger,
+    /**
+     * The dispatcher which will be used for all of the request and response
+     * processing.
+     */
+    dispatcher: CoroutineDispatcher = Dispatchers.IO
+) {
+    private val logger = createChildLogger(parentLogger)
+    private val job = Job()
+    private val coroutineScope = CoroutineScope(dispatcher + job)
+    private val msgsToSend = Channel<Frame>(Channel.RENDEZVOUS)
+    private val msgsReceived = Channel<Frame>(Channel.RENDEZVOUS)
+
+    /**
+     * Send [data] and block until the response is received, returning the
+     * body of the response as a [String].
+     */
+    fun sendAndGetReply(data: String): String {
+        val resp = runBlocking(coroutineScope.coroutineContext) {
+            logger.trace { "sendAndGetReply running in thread ${Thread.currentThread().name}" }
+            msgsToSend.send(Frame.Text(data))
+            msgsReceived.receive().also {
+                logger.trace { "sendAndGetReply got reply in thread ${Thread.currentThread().name}" }
+            }
+        }
+        return when (resp) {
+            is Frame.Text -> resp.readText()
+            else -> throw IllegalStateException("Expected a text frame response")
+        }
+    }
+
+    /**
+     * Send [data] without waiting for its response
+     */
+    fun sendAndForget(data: String) {
+        coroutineScope.launch {
+            logger.trace { "sendAndForget running in thread ${Thread.currentThread().name}" }
+            msgsToSend.send(Frame.Text(data))
+            // Receive the response but ignore it
+            msgsReceived.receive()
+        }
+    }
+
+    /**
+     * Establish the websocket connection and start a loop to handle
+     * sending and receiving websocket messages.
+     */
+    fun run() {
+        coroutineScope.launch {
+            client.ws(host = host, port = port, path = path) {
+                try {
+                    for (msg in msgsToSend) {
+                        send(msg)
+                        val resp = incoming.receive()
+                        msgsReceived.send(resp)
+                    }
+                } finally {
+                    logger.info("Websocket closing")
+                    close()
```

Comment:
What does this do to any pending coroutines?

Reply:
This should clean things up correctly, but I think it'd be good to have a test to verify that...I'll add one.

Reply:
I think the close here was actually redundant, I added some better handling here.